src/main/java/domain/Janggi.java

```diff
@@ -0,0 +1,164 @@
+        UnitType type = pickedUnit.getType();
+        if (type == UnitType.CANNON) {
+            return totalRoutes.stream()
+                    .filter(route -> canCannonJump(pick, route))
+                    .toList();
+        }
+        if (type == UnitType.SOLDIER) {
+            return filterSoldierMoves(pick, pickedUnit, totalRoutes);
+        }
+        return totalRoutes;
+    }
+
+    private List<Route> filterSoldierMoves(Position pick, Unit pickedUnit, List<Route> totalRoutes) {
+        if (pickedUnit.getTeam() == Team.HAN) {
+            return totalRoutes.stream()
+                    .filter(route -> route.getPositions().getFirst().getY() >= pick.getY())
+                    .toList();
+        }
+        return totalRoutes.stream()
+                .filter(route -> route.getPositions().getFirst().getY() <= pick.getY())
+                .toList();
+    }
+
+    private boolean canCannonJump(Position current, Route route) {
+        Position endPoint = route.searchDestination(current);
+        if (!isEmptyPosition(endPoint)) {
+            Unit endUnit = units.get(endPoint);
+            if (endUnit.getType() == UnitType.CANNON) {
+                return false;
+            }
+        }
+
+        int count = 0;
+        for (Position position : route.getPositionsExceptDestination(current)) {
+            if (isEmptyPosition(position)) {
+                continue;
+            }
+            Unit unit = units.get(position);
+            if (unit.getType() == UnitType.CANNON) {
+                return false;
+            }
+            count++;
+        }
+        return (count == 1);
+    }
+
+    private List<Route> filterBlockedRoutes(Position pick, List<Route> routes) {
+        return routes.stream()
+                .filter(route -> isClearRoute(pick, route))
+                .filter(route -> isClearDestination(pick, route))
+                .toList();
+    }
+
+    private boolean isClearRoute(Position pick, Route route) {
+        return route.getPositionsExceptDestination(pick).stream()
+                .allMatch(this::isEmptyPosition);
+    }
+
+    private boolean isClearDestination(Position pick, Route route) {
+        Position endPosition = route.searchDestination(pick);
+        if (isEmptyPosition(endPosition)) {
+            return true;
+        }
+        Unit endPointUnit = units.get(endPosition);
+        return endPointUnit.getTeam() != this.turn;
+    }
+
+    private boolean isEmptyPosition(Position position) {
+        return !units.containsKey(position);
+    }
+
+    public boolean isOneOfTeamNonExist() {
+        long hanUnitCount = units.values().stream().
+                filter(unit -> unit.getTeam() == Team.HAN)
+                .count();
+        long choUnitCount = units.values().stream().
+                filter(unit -> unit.getTeam() == Team.CHO)
+                .count();
+        return (hanUnitCount == 0 || choUnitCount == 0);
+    }
```

Comment:
갯수를 각각 카운트하기보다는 하나라도 존재하는지 여부가 중요한 것으로 보여요. 
그렇다면 아래와 같이 `noneMatch`를 활용하는게 조금 더 의도가 잘 드러나지 않을까요?  😃

```suggestion
    public boolean isOneOfTeamNonExist() {
        boolean isHanUnitNonExist = units.values()
            .stream()
            .noneMatch(unit -> unit.getTeam() == Team.HAN);
        boolean isChoUnitNonExist = units.values().stream()
            .noneMatch(unit -> unit.getTeam() == Team.CHO);
        
        return isHanUnitNonExist || isChoUnitNonExist;
    }
```

Reply:
아니면 아래와 같이 Team을 기준으로 그룹핑해서 결과를 확인해볼 수도 있을 것 같아요. (아예 없는 경우에는 그 Team의 key 값 자체가 없을 것 같아서 key 존재 여부로 판단을 했어요!   더 직관적으로는 `getOrDefault(Team.HAN, 0L)` 같은 방식으로 갯수를 꺼내고 0개인지 확인하는 방식도 있겠지만 로직이 조금 길어지니 가독성이 조금 떨어지는 것 같아서 key 존재 여부 정도로 확인했어요 😃 

```java
        Map<Team, Long> unitCounts = units.values()
            .stream()
            .collect(Collectors.groupingBy(Unit::getTeam, Collectors.counting()));

        return unitCounts.containsKey(Team.HAN) || unitCounts.containsKey(Team.CHO);
```